### PR TITLE
fix for c++17, ubuntu 20.04, gcc 9

### PR DESCRIPTION
**Pull request by longcongduoi**

### CODE REVIEW

**Comment by tvd12** [comment]
@longcongduoi anh có thể paste giúp em lỗi build của anh lên đây được không ạ?

**Comment by longcongduoi** [comment]
[ 50%] Building CXX object CMakeFiles/ezyfox-client.dir/src/manager/EzyAppManager.cpp.o
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘uint16_t com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readUInt16(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:113:2: error: ‘memcpy’ was not declared in this scope
  113 |  memcpy(&pret, p, 2);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:8:1: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
    7 | #include "EzyNull.h"
  +++ |+#include <cstring>
    8 | 
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘uint32_t com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readUInt32(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:120:2: error: ‘memcpy’ was not declared in this scope
  120 |  memcpy(&pret, p, 4);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:120:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘uint64_t com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readUInt64(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:127:2: error: ‘memcpy’ was not declared in this scope
  127 |  memcpy(&pret, p, 8);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:127:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘int16_t com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readInt16(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:138:2: error: ‘memcpy’ was not declared in this scope
  138 |  memcpy(&pret, p, 2);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:138:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘int32_t com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readInt32(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:145:2: error: ‘memcpy’ was not declared in this scope
  145 |  memcpy(&pret, p, 4);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:145:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘int64_t com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readInt64(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:152:2: error: ‘memcpy’ was not declared in this scope
  152 |  memcpy(&pret, p, 8);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:152:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘float com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readFloat(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:159:2: error: ‘memcpy’ was not declared in this scope
  159 |  memcpy(&pret, p, 4);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:159:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp: In member function ‘double com::tvd12::ezyfoxserver::client::codec::EzyDataDecoder::readDouble(const char*)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:166:2: error: ‘memcpy’ was not declared in this scope
  166 |  memcpy(&pret, p, 8);
      |  ^~~~~~
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/codec/EzyDataDecoder.cpp:166:2: note: ‘memcpy’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
[ 52%] Building CXX object CMakeFiles/ezyfox-client.dir/src/manager/EzyHandlerManager.cpp.o
make[2]: *** [CMakeFiles/ezyfox-client.dir/build.make:370: CMakeFiles/ezyfox-client.dir/src/codec/EzyDataDecoder.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....

**Comment by longcongduoi** [comment]
[  1%] Building CXX object CMakeFiles/ezyfox-client.dir/src/manager/EzyAppManager.cpp.o
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/manager/EzyAppManager.cpp: In member function ‘com::tvd12::ezyfoxserver::client::entity::EzyApp* com::tvd12::ezyfoxserver::client::manager::EzyAppManager::removeApp(int)’:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/manager/EzyAppManager.cpp:23:72: error: no matching function for call to ‘find(std::vector<com::tvd12::ezyfoxserver::client::entity::EzyApp*>::iterator, std::vector<com::tvd12::ezyfoxserver::client::entity::EzyApp*>::iterator, com::tvd12::ezyfoxserver::client::entity::EzyApp*&)’
   23 |         auto appIndex = std::find(mAppList.begin(), mAppList.end(), app);
      |                                                                        ^
In file included from /usr/include/c++/9/bits/locale_facets.h:48,
                 from /usr/include/c++/9/bits/basic_ios.h:37,
                 from /usr/include/c++/9/ios:44,
                 from /usr/include/c++/9/istream:38,
                 from /usr/include/c++/9/sstream:38,
                 from /home/longdh/working/easyfox/ezyfox-server-cpp-client/src/manager/../entity/EzyValue.h:6,
                 from /home/longdh/working/easyfox/ezyfox-server-cpp-client/src/manager/../entity/EzyApp.h:6,
                 from /home/longdh/working/easyfox/ezyfox-server-cpp-client/src/manager/EzyAppManager.cpp:3:
/usr/include/c++/9/bits/streambuf_iterator.h:373:5: note: candidate: ‘template<class _CharT2> typename __gnu_cxx::__enable_if<std::__is_char<_CharT2>::__value, std::istreambuf_iterator<_CharT> >::__type std::find(std::istreambuf_iterator<_CharT>, std::istreambuf_iterator<_CharT>, const _CharT2&)’
  373 |     find(istreambuf_iterator<_CharT> __first,
      |     ^~~~
/usr/include/c++/9/bits/streambuf_iterator.h:373:5: note:   template argument deduction/substitution failed:
/home/longdh/working/easyfox/ezyfox-server-cpp-client/src/manager/EzyAppManager.cpp:23:72: note:   ‘__gnu_cxx::__normal_iterator<com::tvd12::ezyfoxserver::client::entity::EzyApp**, std::vector<com::tvd12::ezyfoxserver::client::entity::EzyApp*> >’ is not derived from ‘std::istreambuf_iterator<_CharT>’
   23 |         auto appIndex = std::find(mAppList.begin(), mAppList.end(), app);
      |                                                                        ^
make[2]: *** [CMakeFiles/ezyfox-client.dir/build.make:454: CMakeFiles/ezyfox-client.dir/src/manager/EzyAppManager.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/ezyfox-client.dir/all] Error 2
make: *** [Makefile:91: all] Error 2

**Comment by longcongduoi** [comment]
> @longcongduoi anh có thể paste giúp em lỗi build của anh lên đây được không ạ?

Tham khảo chỗ này https://en.cppreference.com/w/cpp/algorithm/find
thì hàm find em đang dùng chỉ có trên c++20 thôi

File: src/gc/EzyAutoReleasePool.cpp [hunk 1]
```diff
@@ -1,4 +1,6 @@
 #include <thread>
+#include <mutex>
```
Comment (tvd12) [inline src/gc/EzyAutoReleasePool.cpp:2]: LGTM

File: src/codec/EzyDataDecoder.cpp [hunk 2]
```diff
@@ -6,6 +6,9 @@
 #include "EzyByteArray.h"
 #include "EzyNull.h"
 
+// FIXME
+#include <cstring>
```
Comment (tvd12) [inline src/codec/EzyDataDecoder.cpp:10]: Anh có thể bỏ `// FIXME` và chuyển `#include <cstring>` lên đầu được không ạ?

File: src/manager/EzyAppManager.cpp [hunk 3]
```diff
@@ -5,6 +5,21 @@
...
+{
+    for (; first != last; ++first) {
+        if (*first == value) {
+            return first;
+        }
+    }
+    return last;
+}
+#endif
+
```
Comment (tvd12) [inline src/manager/EzyAppManager.cpp:22]: hiện tại em mới test trên C++ 11 và C++ 14 với gcc, nên khả năng nó sẽ còn nhiều lỗi cho các trình dịch khác nên em muốn sửa nó thành:

```
namespace stdex {
template<class InputIterator, class T>
InputIterator find (InputIterator first, InputIterator last, const T& val) {
    while (first!=last) {
        if (*first==val) {
            return first;
        }
        ++first;
    }
    return last;
}
}
```

và ở dưới sửa thành: `stdex::find`

Anh thấy sao ạ?

**Review by tvd12 — CHANGES_REQUESTED** [review]
@longcongduoi PLTAL

File: src/manager/EzyAppManager.cpp (hunk 3 above)
Comment (longcongduoi) [reply src/manager/EzyAppManager.cpp:22]: Cách của em hay hơn đấy

File: src/codec/EzyDataDecoder.cpp (hunk 2 above)
Comment (longcongduoi) [reply src/codec/EzyDataDecoder.cpp:10]: Sorry em, anh thêm dòng này vô để compile test cho nhanh (nên anh thêm chữ fixme vô, ko chắc cách này tối ưu)

File: src/gc/EzyAutoReleasePool.cpp (hunk 1 above)
Comment (longcongduoi) [reply src/gc/EzyAutoReleasePool.cpp:2]: Bên anh nó báo lỗi mutex và khuyến nghị thêm #include <mutex> vô, anh cũng chưa check lại

Comment (tvd12) [reply src/gc/EzyAutoReleasePool.cpp:2]: Vâng, em dùng xcode với cả android ndk thì ko thấy báo gì, em nghĩ thêm vào cũng không sao anh ạ

File: src/manager/EzyAppManager.cpp [hunk 4]
```diff
@@ -5,6 +5,23 @@
 
 EZY_NAMESPACE_START_WITH(manager)
 
+using namespace std;
+
+#if (cplusplus <= 201703L)
```
Comment (tvd12) [inline src/manager/EzyAppManager.cpp:10]: Ý em là mình bỏ hẳn cái này `#if (cplusplus <= 201703L)` đi ạ, vì nếu vẫn còn check thế này thì build cho những loại khác sẽ failed ạ

File: src/manager/EzyAppManager.cpp [hunk 5]
```diff
@@ -5,6 +5,23 @@
...
+namespace stdex {
+    template<class InputIt, class T>
+    constexpr InputIt find(InputIt first, InputIt last, const T& value)
+    {
```
Comment (tvd12) [inline src/manager/EzyAppManager.cpp:14]: coding convention của mình là dấu `{` ở ngay sau khai báo hàm kiểu `constexpr InputIt find(InputIt first, InputIt last, const T& value) {` anh có thể giúp em sửa được không ạ?

File: src/manager/EzyAppManager.cpp [hunk 6]
```diff
@@ -5,6 +5,23 @@
...
+        for (; first != last; ++first) {
```
Comment (tvd12) [inline src/manager/EzyAppManager.cpp:15]: Em tham khảo thì mình không nên gán lại giá trị của paramter: https://ncrcoe.gitbooks.io/java-for-small-teams/content/specifics/200_do_not_reassign_parameters.html
Em nghĩ tốt hơn sẽ là:

```cpp
namespace stdex {
template<class InputIt, class T>
constexpr InputIt find(InputIt first, InputIt last, const T& value){
    for (InputIt answer = first; answer != last; ++answer) {
        if (*answer == value) {
            return first;
        }
    }
    return last;
}
}
```

anh thấy sao ạ?

**Review by tvd12 — CHANGES_REQUESTED** [review]
PLTAL

**Review by longcongduoi — COMMENTED** [review]
Okie em, để anh sửa nha

File: src/manager/EzyAppManager.cpp (hunk 5 above)
Comment (longcongduoi) [reply src/manager/EzyAppManager.cpp:14]: Để anh sửa nha, sorry em là anh đọc tài liệu chưa kỹ

**Comment by tvd12** [comment]
@longcongduoi bọn em cũng định hướng dùng cho các hệ thống IoT nữa ạ

**Comment by tvd12** [comment]
@longcongduoi còn cái này anh thấy có cần thiết không ạ? https://github.com/youngmonkeys/ezyfox-server-cpp-client/pull/20#discussion_r684595501

File: src/manager/EzyAppManager.cpp (hunk 6 above)
Comment (longcongduoi) [reply src/manager/EzyAppManager.cpp]: Anh thấy dùng biến Iter mới sẽ trong sáng/rõ ràng hơn, đẻ anh fix lại chỗ này theo suggest của em

**Review by tvd12 — APPROVED** [review]
@longcongduoi LGTM, cám ơn anh nhiều ạ

**Pull request merged**